### ISSUE TITLE
How to deal with preferences with multiple environments and multiple napari versions

### ISSUE THREAD

**Issue by jni** [napari/napari#4697]
### Overview of design need

Preferences are a slightly weird thing in Python, because users typically have more than one Python environment in their machine, and napari is constantly evolving (so some preferences for 0.4.15 might not make sense for 0.4.16, and vice versa, but both versions might exist alongside each other in different environments). Additionally, users might have different preferences when using napari as a standalone application compared to using it from Python. Finally, we store preferences related to plugins, and not all plugins are present in all environments.

We discussed some ideas for this in the eurasia community meeting. None seem like a silver bullet, but here are the suggestions:

- [x] create different preference files keyed by environment/interpreter. So, if I conda activate new-environment, I get a clean napari. Optionally, we could offer to import/migrate preferences from prior environments on first launch.
- better support for exporting and importing preferences, both from the command line and from the UI, would alleviate this problem, as users could have a bunch of preference files that they self-manage, e.g. in Dropbox.
- cascading preferences, so that napari would look for preferences in system, user, version, and environment-specific folders, and load them each in turn, clobbering any conflicting settings, so that preferences could be increasingly specific.

CC @haesleinhuepf @Czaki @jaimergp @DragaDoncila, who participated in discussions and offered some of the above suggestions. Overall, personally, I like per-env plus import from previous env the best.

### What level of design is needed? (Choose all that apply)

- [x] **User experience research:**  high-level recommendation/exploration of user needs, design heuristics, and / or best practices to inform a design experience  
(Use this option when you feel there’s a challenge to be solved, but you’re curious about what the experience should be — may involve research studies to understand challenges/opportunities for design)

### Is design a blocker?
- [ ] **Yes:** engineering cannot proceed without a design first
- [ ] **No:** engineering can create a first version, and design can come in later to iterate and refine
- [x] 🤷 dunno could go either way really 😂

**Comment by tlambert03** [comment]
> cascading preferences, so that napari would look for preferences in system, user, version, and environment-specific folders, and load them each in turn, clobbering any conflicting settings, so that preferences could be increasingly specific.

I like this.  
I don't love the user interaction required in the "would you like to import from previous environment?"... and it seems like something that could break a lot given the heterogeneity of environments.

how does ipython and or matplotlib deal with this?

**Comment by jaimergp** [comment]
From [IPython docs](https://ipython.readthedocs.io/en/stable/config/intro.html):

>  Configuration files are loaded in order and merged with configuration on later location taking precedence on earlier locations (that is to say a user can overwrite a system wide configuration option).
> 
> You can see all locations in which IPython is looking for configuration files by starting ipython in debug mode:
>
> ```
> $ ipython --debug -c 'exit()'
> ```

They also have a `--profile` flag to load a different configuration.

**Comment by DragaDoncila** [comment]
>  Optionally, we could offer to import/migrate preferences from prior environments on first launch

I'm super hesitant on this because I spin up bulk environments to test stuff and this would be annoying. Yes, we could add a preference for whether we ask or not, but I'm also very conscious of preference creep in general.

> better support for exporting and importing preferences, both from the command line and from the UI, would alleviate this problem, as users could have a bunch of preference files that they self-manage, e.g. in Dropbox.

I think this would be ok, and should probably be done in conjunction with other methods of inferring the correct preference file e.g. the hierarchical approach you're proposing. You should always be able (via the command line or the GUI) to override your preferences by pointing to a different file.

> cascading preferences, so that napari would look for preferences in system, user, version, and environment-specific folders, and load them each in turn, clobbering any conflicting settings, so that preferences could be increasingly specific.

I feel like I'd want us to expand on what this looks like a bit before supporting it. In this cascade world, where do preferences I change go? In the most specific bucket i.e. environment? Can I float preferences up the chain? How do I specify where a given preference is loaded from? Do we need a preference file for preferences?

I've often thought that plugin preferences should be environment specific by default. I could always point to a different file if I desperately wanted to (via option 1 above), but it's hard for me to imagine wanting e.g. reader preferences to be loaded when I spin up a new environment because it's just extremely unlikely they'll make sense. Maybe it's also because I've been testing file opening preferences a lot, but yeah, it was annoying having to clear them all the time.

**Comment by jaimergp** [comment]
I recently rediscovered the [docs for conda config](https://docs.conda.io/projects/conda/en/latest/user-guide/configuration/use-condarc.html#searching-for-condarc), which might be relevant in terms of where configuration files are searched, the precedence and how conflicting values are merged.

**Comment by psobolewskiPhD** [comment]
Going to bump this as I ran into it recently where installing `main` which changed keybinding mechanism resulted in an error in my live 0.4.16 working napari env. `napari --reset` fixed everything but it's less than ideal.
I'm a big fan of all of these options from the OP:

> We discussed some ideas for this in the eurasia community meeting. None seem like a silver bullet, but here are the suggestions:
> 
> * create different preference files keyed by environment/interpreter. So, if I conda activate new-environment, I get a clean napari. Optionally, we could offer to import/migrate preferences from prior environments on first launch.
> * better support for exporting and importing preferences, both from the command line and from the UI, would alleviate this problem, as users could have a bunch of preference files that they self-manage, e.g. in Dropbox.
> * cascading preferences, so that napari would look for preferences in system, user, version, and environment-specific folders, and load them each in turn, clobbering any conflicting settings, so that preferences could be increasingly specific.

I think it's very much worth exploring how to better handle settings between conda envs and how to permit users to import/export settings more easily from the GUI.

In the meantime, perhaps making settings file version specific would be a good step forward?
For example, the `settings.yaml` could be `settings_v<version>.yaml`? Then napari v<version> would look for `settings_v<version>.yaml` and fall back to `settings.yaml` if it's missing.

**Comment by Czaki** [comment]
> For example, the `settings.yaml` could be `settings_v<version>.yaml`

I rather suggest having a separate directory per the napari version.

**Comment by psobolewskiPhD** [comment]
> > For example, the `settings.yaml` could be `settings_v<version>.yaml`
> 
> I rather suggest having a separate directory per the napari version.

So in the settings path `napari` folder would be replaced with `napari_v<version>` and the file would still be `settings.yaml`? 
Or nested one more level: `napari/v<version>/settings.yaml`
Either option seems reasonable to me, because it would make moving settings around easier.

**Comment by Czaki** [comment]
> Or nested one more level: `napari/v<version>/settings.yaml`

this one is best for me. 

y motivation is that my plugin store its settings in the same directory as the current napari instance settings.

**Comment by goanpeca** [comment]
Agreed with @Czaki suggestion on using folders!

I can take a look at this one next week. Sorry for the issues with your preferences with 0.4.16 @psobolewskiPhD !

We could start by fixing preferences by versions, now to deal with environments, we could cascade

`Current working directory > Current environment > User by version > System by version`

**Comment by psobolewskiPhD** [comment]
♥️ @goanpeca 
I will also poke around maybe can get somewhere with the folders idea, if I have some time this week.

**Comment by psobolewskiPhD** [comment]
So the version-subfolder approach is actually really easy to implement because napari uses `appdirs` (HT @Czaki)
https://github.com/napari/napari/blob/3779997fbfa23f9e02bf29f5087f857e8246b846/napari/utils/_appdirs.py#L16-L35

Now I'm not sure I totally follow the comment.
I think it's referring to the fact that napari `v<n>` will only look for config in folder `v<n>`, so if you update from `v<n-1>` then you have a problem.
I think this could be solved though with checking if the path exists and if not importing from folder `v<n-1>`, if that exists, or just folder napari if that exists.

But where does that kind of logic belong best?

For importing, at the moment, a variable `_DEFAULT_CONFIG_PATH` is set here:
https://github.com/napari/napari/blob/5f96d5d814aad697c367bdadbb1a57750e2114ad/napari/utils/_base.py
It's then used in two places. Settings:
https://github.com/napari/napari/blob/ead5e5051aa0bb850b52e72cfed2bcd5c26b877b/napari/settings/_napari_settings.py
and in translations, where a check if the path exists is already being done:
https://github.com/napari/napari/blob/1f02839ddf9b1beef41d31b43bc7cec29c00d867/napari/utils/translations.py#L657-L681

**Comment by Czaki** [comment]
For me, the best option is if the directory `vX.Y.Z` is not available then the most recent previous version should be copied. Ten after the update user will have his settings, but when downgraded, then the previous version still exists. 

I use this approaching PartSeg, and it works for me. 

Here the code for coping that I already use: https://github.com/4DNucleome/PartSeg/blob/42b1d71ed7d70ffb81adfbb4285dc6673f85d50d/package/PartSeg/common_backend/load_backup.py

**Comment by psobolewskiPhD** [comment]
Yeah, that makes sense and what I was trying to say—my post above had some formatting issue so it didn't come out that way.
So where would such checks live in the napari codebase?

**Comment by jaimergp** [comment]
We are keying by version assuming there might be changes in every release. I think we should be instead generating a versioned schema of the config file, with minimum napari release requirements or something like that. An import mechanism should be able to "rescue" old files an produce the modern schema-compliant mechanism. This would also allow us to offer compatibility hooks on newer naparis, to a reasonable extent.

On top of that schema approach I would then implement the cascading bits.

Does that make sense? I didn't think too much about it so I might be missing details, but versioned schemas sound like a good idea.

**Comment by psobolewskiPhD** [comment]
To give an update here:
Right now, the issue of `main` or `dev` vs. live napari is resolved, because `dev` settings have been silo'd, the PR is merged: #5322
So when using a `dev` release, a new `config_path` is set and no attempt is made to import previous settings from a different path nor make dev settings available to a non-dev, live napari. Switching envs
There is also an open PR to make settings versioned, using subdirectories: #5306 
This PR would include release versions and does include a mechanism of looking for directories of previous versions.

Today we (me, @Czaki and @jaimergp) discussed this issue a bit at the Community meeting today.
One thing that came up is the case of the same napari version but in different envs—this should probably be distinct. 
Another thing discussed was that when an existing napari is updated, then for sure settings need to be migrated if possible. But if a user makes a fresh env, they really should get a fresh napari.
So @jaimergp  and @Czaki came up with a clever strategy to use as the settings sub-dir a hash the location of the python executable `sys.executable`. This means that settings would be on a per env basis, with napari in one env not looking in another, but at the same time, changing napari in that env would preserve settings—schema permitting.
Importantly, this solution would also nicely handle plugin settings, assuming that plugins would be instructed to use a path within the napari `config_path`.
(A second layer to consider is providing the ability to import from another env.)
In summary, I think this is a very good solution to both of the use case issues mentioned above.
There is an initial implementation here https://github.com/napari/napari/pull/5333 by @Czaki.
(Note that for me on macOS lauching napari from terminal makes a new path each time, but from ipython it works correctly).

**Comment by psobolewskiPhD** [comment]
I took the liberty of editing the OP to mark that the first point (per env preferences) is solved, via https://github.com/napari/napari/pull/5333
The question of import/export of preferences and cascading remains, so this should probably stay open.
